Manage column formatting for datasets + add extra form fields (UI)

kulyk opened the pull request:
Adds a few missing dataset field metadata form fields:

* "Special Type" picker (a.k.a field semantic type)
* "Column Formatting" section
* mapped database column field for native datasets (people can select to which real DB column a dataset field is mapped to (this can unlock column-level permission management, sandboxing, etc. on datasets)

⚠️  "Special Type" and "Mapped DB column" pickers need a BE API available to function correctly, so for now there is only a UI for them. Editing these fields won't be persisted or change something on the UI. On the other hand, the "Column Formatting" relies on the card's `visualization_settings`, so it should work properly and you should see immediate feedback when changing formatting options.

### To Verify

1. Open a GUI dataset (they usually have more rich metadata, so you'd see more different states on the UI)
2. Click the dataset name to open the details sidebar and click the "Customize metadata" button
3. Click through a few columns. In the sidebar, you should now see a column formatting section. E.g. for numeric columns, you should see "show progress mini-bar", "number separator" options, etc.
4. Try editing a few "column formatting" options. You should see immediate changes in the results table. Click "Save changes", refresh the page. Changes to column formatting should be persisted
5. Click the "Special type" picker in the field metadata sidebar. It should now expand instead of being a dropdown list. You should also be able to search for a particularly special type by name. ⚠️ This field requires the BE API available and changing it won't currently do anything and the value won't be saved on the server.
6. For a few fields, try selecting a Currency special type (e.g. Cost). A currency picker should appear right behind the special type picker
7. For a few fields, try selecting an FK special type. An FK target picker should appear right behind the special type picker (you should be able to pick to which table an FK corresponds to)
8. Open a native dataset picker. For every field, you should see a "Database column this maps to" field. Click on the select, you should see a picker with a list of database tables. Click on a table, you should see a list of table columns. Click any column, the select should display "TABLE_NAME > FIELD_NAME". ⚠️ This field requires the BE API available and changing it won't currently do anything and the value won't be saved on the server.

### Demo

<details>
<summary>UI + Column Formatting Changes</summary>

https://user-images.githubusercontent.com/17258145/145438909-b0a77fbb-fa37-454c-b6b3-f27b6903be77.mov
</details>

<details>
<summary>Native datasets' "Mapped DB column" field UI demo </summary>

https://user-images.githubusercontent.com/17258145/145439096-997ccd7e-27d2-4f25-a327-3ae2645b1ba7.mov

</details>

kulyk committed:
Convert SelectButton to TypeScript

kulyk committed:
Add onClick prop to SelectButton

kulyk committed:
Extract FormFieldDivider component

kulyk committed:
Pass dataset into field metadata sidebar

kulyk committed:
Add function to get field metadata form fields

kulyk committed:
Simplify field visibility options generation

kulyk committed:
Allow to provide custom form field widgets

kulyk committed:
Add `renderSearchSection` prop to AccordionList

kulyk committed:
Add SemanticTypePicker component

kulyk committed:
Use SemanticTypePicker

kulyk committed:
Add FieldSelector (DataSelector subset)

kulyk committed:
Allow hiding DataSelector trigger expand control

kulyk committed:
Add MappedFieldPicker component

kulyk committed:
Use MappedFieldPicker for native datasets

kulyk committed:
Simplify mapped field name formatting

kulyk committed:
Fix dataset type usage

kulyk committed:
Remove redundant props

kulyk committed:
Disable fuzzy search for SemanticTypePicker

kulyk committed:
Allow hiding empty AccordionList sections in search

kulyk committed:
Don't show empty sections in SemanticTypePicker

kulyk committed:
Add formatting section to field metadata form

kulyk committed:
Fix FormFieldDivider margin

kulyk committed:
Add FK target and currency pickers

kulyk committed:
Hide column title from formatting section

kulyk committed:
Wire up column formatting updates

kulyk committed:
Convert ChartSettingsWidget to TypeScript

kulyk committed:
Use styled-components for ChartSettingsWidget

kulyk committed:
Add "variant" prop to ChartSettingsWidget

kulyk committed:
Unify sidebar form UI

kulyk committed:
Fix hidden column settings are forcefully shown

mazameli commented:
I know my designs didn't show a field label for the column type select, but seeing it in the branch now, I think that at least in its collapsed state it could use a `Column type` label.

![image](https://user-images.githubusercontent.com/2223916/145646286-257779cc-5b09-4a56-a3cc-9599ff18f0e8.png)

mazameli commented:
I think that we're going to need to add tabs for **date** and **numeric** columns to separate all their different additional options into `Settings` and `Formatting` tabs, like in this mockup. All other field types (categorical ones, etc.) can stay the way they already are in the branch since they don't have any formatting options. Hopufully this isn't a rough change, but let me know what you think.

![image](https://user-images.githubusercontent.com/2223916/145648469-543a1483-be63-4ed9-b4a6-6b2048ae3b45.png)

mazameli commented:
The form labels in the Simple mode sidebar somehow got too large in this branch (not sure it was first caused in this branch or elsewhere). On the left is this branch (labels too big), on the right is master (correct label sizing):

![image](https://user-images.githubusercontent.com/2223916/145649663-ba6e9101-fc73-40b2-a011-93353d42cf34.png)

mazameli commented:
This is probably an error due to me not calling this out explicitly (apologies), but `Display as` is _not_ meant to be a _new_ option: I intended to _change_ the existing `View as link or image` setting to this style, i.e. by:
- changing the setting's label
- -changing it from a dropdown to radio buttons
- changing the `Off` option to `Text`
- and doing both of the above wherever this setting is accessible, since I assume it's a shared component (in the viz settings side panel, and in the column settings in the admin panel data model)

![image](https://user-images.githubusercontent.com/2223916/145650129-14a1cad7-7b0e-4907-81ad-99124541182e.png)

kulyk committed:
Add "column type" field label

kulyk committed:
Fix ChartSettingsWidget label size in defaul mode

kulyk committed:
Use "radio" input for "view_as" field

kulyk committed:
Use "Text" instead of "Off" for view_as field

kulyk committed:
Use tab layout in field metadata sidebar

kulyk committed:
Hide Formatting tab if missing formatting opts

kulyk committed:
Fix hidden formatting fields behavior

kulyk commented:
Thanks, @mazameli 🙏 Could you please take one more look at the updated changes?

1. Added the "Column type" form field label
2. Added "Settings" and "Formatting" tabs for the field sidebar. Looks much cleaner to me. I also made it hide the "Formatting" tab if there are no formatting options for a selected "special type", does it make sense to you? Heads-up: changing the "special type" field value on the UI won't show/hide the "Formatting" tab in this case since it's not connected to the BE yet
3. Fixed form label sizes in simple mode sidebar. Messed that up while extending its UI for the dataset editor sidebar. Great catch!
4. Fixed the "Display as text/link" field. Removed it from the "Formatting" section and updated the form widget to always use the radio input. Thanks for catching that!

I also have two questions here:

1. Did you test the expanded state for the "special type" picker? I'm not sure about the inner scroll experience here, would appreciate your feedback there
2. On the "Display as text/link" field. If a column is missing the special type, we offer a few other options like "Email link", "Image", "Automatic", so the list of options becomes longer (see the screenshot below). Do we still want to use the "radio" control in this case? As an option, we can choose the widget depending on the options count. So we can show the radio control for "Text" / "Link" options and use the dropdown select if we have > 2 options


<img width="334" alt="CleanShot 2021-12-13 at 18 03 36@2x" src="https://user-images.githubusercontent.com/17258145/145846620-7326f536-d15b-461b-8e38-d1f3eccefeed.png">

kulyk committed:
Fix E2E test using renamed label

daltojohnso reviewed (approved):
Looks good -- just one minor thing

mazameli committed:
change view_as label to "Display as"

mazameli commented:
All the form input selection values seem to be `text-medium` instead of the expected `text-dark`. 

![image](https://user-images.githubusercontent.com/2223916/145906329-3b319798-2376-4f83-96c7-5428206b7f97.png)

mazameli commented:
> Did you test the expanded state for the "special type" picker? I'm not sure about the inner scroll experience here, would appreciate your feedback there.

I think the problem is that it appear to have a fixed height when it's expanded, so that you have to scroll the viewport in order to see the bottom options, which feels wrong. I don't know if it's possible, but I think what you'd want instead is to give it a minimum height that's much smaller (like 350px) but have the height itself be responsive and extend to the bottom of the viewport, so that you could make sure you're able to scroll to the last option inside that element without having to scroll the page.  

Let me know if I'm making any sense. :b

mazameli commented:
Now that we have a form label for this input in this state, we can remove this divider line/rule:

![image](https://user-images.githubusercontent.com/2223916/145907201-8a7c551f-56b1-4059-8921-b75fd50fd504.png)

mazameli commented:
> So we can show the radio control for "Text" / "Link" options and use the dropdown select if we have > 2 options

Love it, let's do that.

mazameli commented:
This is coming along nicely, @kulyk! Ping me if you need to work any of this out together in real-time.

kulyk committed:
Remove divider from expanded semantic type picker

kulyk committed:
Move FormFieldDivider into sidebar's .styled file

kulyk committed:
Choose "view_as" widget depending on options count

* use "radio" for exactly two options
* use "select" when there are > 2 options

kulyk committed:
Change selected radio button look

kulyk committed:
Use "text-dark" in sidebar form

kulyk committed:
Set height for semantic type picker

kulyk committed:
Use Field's displayName method for formatting

kulyk commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetFieldMetadataSidebar/MappedFieldPicker/MappedFieldPicker.tsx:
```diff
@@ -0,0 +1,49 @@
+  selectedField?: Field;
+};
+
+function formatFieldLabel(field: Field) {
```
Fixed! Thank's for pointing that out

mazameli reviewed (approved):
Nice work!
